thread.c

```diff
@@ -189,6 +189,57 @@ void pause_threads(enum pause_thread_types type) {
     pthread_mutex_unlock(&init_lock);
 }
 
+// MUST not be called with any deeper locks held
+// MUST be called only by parent thread
+// Note: listener thread is the "main" event base, which has exited its
+// loop in order to call this function.
+void stop_threads(void) {
```

Comment:
Ideally you'd stop the threads in the reverse order in which they're started, but due to the interactions between these threads stopping the assoc maintenance thread first seems like the right thing to do to avoid races between `pause_threads()` and the remaining part of `stop_threads()`.

An open question which I haven't been able to answer yet is whether the worker threads, the crawler thread and the LRU-maintainer thread, expect the assoc maintenance thread to "do it's job" in any edge case between the time we stop the assoc maintenance thread and stop the remaining threads.


This is the thread start order from my reading of the code (if this helps you at all):
1. Logger thread
2. Init worker threads
3. Assoc maintenance thread
4. Item crawler thread
5. Storage compact thread (EXTSTORE)
6. Storage write thread (EXTSTORE)
7. LRU Maintainer thread
8. Slab maintenance thread 
9. Conn timeout thread

Reply:
Thanks a lot for looking this part over. I think this just sucks for now, without proper RCU of the hash table... assoc has to pause everything to flip that pointer around. Perhaps with my cleanup of the worker thread situation, I can put the hash table into the thread struct and use a proper signal -> check-in routine. That'd be great to fix since it does briefly halt all operations while flipping that pointer. Over the last decade that routine has evolved into a shorter and shorter pause :)

It should be able to bail and shut off early if it's using worker pipe signals, since it won't care about completing its hash expansion if we're shutting down anyway.

Reply:
Sounds great, as I get more acquainted with the code base, I'll try to see if I can find any potential races. As for avoiding the pause, that'd be great to have.